docs/source/advanced/backends.rst

```diff
@@ -1,14 +1,108 @@
+
+Fibers
+``````
+
+Execution
++++++++++
+
+To prevent recreation of the fibers between execution of different blocks in the grid, the fibers are stored inside a fibers pool.
+This fiber pool is local to the invocation because making it local to the KernelExecutor could mean a heavy memory usage when there are multiple KernelExecutors around.
+
+OpenMP
+``````
+
+Execution
++++++++++
+
+Parallel execution of the kernels in a block is required because when syncBlockThreads is called all of them have to be done with their work up to this line.
+So we have to spawn one real thread per kernel in a block.
+``omp for`` is not useful because it is meant for cases where multiple iterations are executed by one thread but in our case a 1:1 mapping is required.
+Therefore we use ``omp parallel`` with the specified number of threads in a block.
+Another reason for not using ``omp for`` like ``#pragma omp parallel for collapse(3) num_threads(blockDim.x*blockDim.y*blockDim.z)`` is that ``#pragma omp barrier`` used for intra block synchronization is not allowed inside ``omp for`` blocks.
+
+Because OpenMP is designed for a 1:1 abstraction of hardware to software threads, the block size is restricted by the number of OpenMP threads allowed by the runtime.
+This could be as little as 2 or 4 kernels but on a system with 4 cores and hyper-threading OpenMP can also allow 64 threads.
+
+Index
++++++
+
+OpenMP only provides a linear thread index. This index is converted to a 3 dimensional index at runtime.
+
+Atomic
+++++++
+
+We can not use '#pragma omp atomic' because braces or calling other functions directly after ``#pragma omp atomic`` are not allowed.
+Because we are implementing the CUDA atomic operations which return the old value, this requires ``#pragma omp critical`` to be used.
+``omp_set_lock`` is an alternative but is usually slower.
+
+CUDA
+````
 
-The element level is supported through an additional run-time variable containing the extent of elements per thread.
-This variable can be accessed by all threads and should optimally be placed in constant device memory for fast access.
+Nearly all CUDA functionality can be directly mapped to alpaka function calls.
+A major difference is that CUDA requires the block and grid sizes to be given in (x, y, z) order.
+alpaka uses the mathematical C/C++ array indexing scheme [z][y][x].
+Dimension 0 in this case is z, dimensions 2 is x.
 
-Porting CUDA to *alpaka*
-------------------------
+Furthermore alpaka does not require the indices and extents to be 3-dimensional.
+The accelerators are templatized on and support arbitrary dimensionality.
+NOTE: Currently the CUDA implementation is restricted to a maximum of 3 dimensions!
+
+NOTE: The CUDA-accelerator back-end can change the current CUDA device and will NOT set the device back to the one prior to the invocation of the alpaka function!
```

Comment:
This part is mostly a copy of the next section, from line 110. I think it should be removed either here or there.